### PR TITLE
Add support for credsStore/credHelpers

### PR DESCRIPTION
**Pull request by evelynl94**

### CODE REVIEW

**Comment by evelynl94** [comment]
To add the helper binaries (`amazon-ecr-credential-helper` for example) in Makisu's Dockerfile
```
FROM golang:1.11 AS builder

RUN mkdir -p /go/src/github.com/uber/makisu
WORKDIR /go/src/github.com/uber/makisu

ADD Makefile .
RUN make ext-tools/Linux/dep

ADD Gopkg.toml Gopkg.lock ./
ADD .git ./.git
ADD cli ./cli
ADD bin ./bin
ADD lib ./lib
RUN make bins

# Build Amazon ECR credential helper
RUN go get -u github.com/awslabs/amazon-ecr-credential-helper/ecr-login/cli/docker-credential-ecr-login
RUN make -C /go/src/github.com/awslabs/amazon-ecr-credential-helper linux-amd64

FROM scratch
COPY --from=builder /go/src/github.com/uber/makisu/bin/makisu/makisu /makisu-internal/makisu
COPY --from=builder /go/src/github.com/awslabs/amazon-ecr-credential-helper/bin/linux-amd64/docker-credential-ecr-login /makisu/docker-credential-ecr-login
ADD ./assets/cacerts.pem /makisu-internal/certs/cacerts.pem
ENTRYPOINT ["/makisu-internal/makisu"]
```
And pass in the environment variables `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` with registry config: 
```
"someawsregistry":
  ".*":
    security:
      credStore: ecr-login
```

**Comment by geekflyer** [comment]
@evelynl94 First of all thanks for implementing this! 

I'm trying to use this new cred store feature now in conjunction but running into a few issues.
1. your example above contains a typo in the registry config. It should be `credsStore` instead of `credStore`. Took me a bunch of try-and-error and digging into the source code to find that out. Consider this problem solved for me but perhaps it would be good if you update the example above.

2. I built an image which contains makisu and the gcr credential helper using the following dockerfile:
```
FROM golang:1.11 AS builder

RUN mkdir -p /go/src/github.com/uber/makisu
WORKDIR /go/src/github.com/uber/makisu

ADD Makefile .
RUN make ext-tools/Linux/dep

ADD Gopkg.toml Gopkg.lock ./
ADD .git ./.git
ADD cli ./cli
ADD bin ./bin
ADD lib ./lib
RUN make bins

# Build GCR credential helper
RUN go get -u github.com/GoogleCloudPlatform/docker-credential-gcr
RUN make -C /go/src/github.com/GoogleCloudPlatform/docker-credential-gcr

FROM scratch
COPY --from=builder /go/src/github.com/uber/makisu/bin/makisu/makisu /makisu-internal/makisu
COPY --from=builder /go/src/github.com/GoogleCloudPlatform/docker-credential-gcr/bin/docker-credential-gcr /makisu-internal/docker-credential-gcr
ADD ./assets/cacerts.pem /makisu-internal/certs/cacerts.pem
ENTRYPOINT ["/makisu-internal/makisu"]
```
and using the command `docker build -f makisu-gcr.Dockerfile . -t makisu-gcr`

Then I attempted to use the new image to built my application via:
```
docker run -i --rm --network host \
        -v /var/run/docker.sock:/docker.sock \
        -v $HOME/.config/gcloud:/root/.config/gcloud -e DOCKER_HOST=unix:///docker.sock \
        -v $(pwd):/makisu-context \
        -v /tmp/makisu-storage:/makisu-storage \
        makisu-gcr:latest build --modifyfs=true --load -registry-config=/makisu-context/registry.yaml -t=gcr.io/my-project/christian-playground/makisu --push=gcr.io -f /makisu-context/test.Dockerfile /makisu-context
```

However makisu build fails with:

```
➜  makisu-use docker run -i --rm --network host \
        -v /var/run/docker.sock:/docker.sock \
        -v $HOME/.config/gcloud:/root/.config/gcloud -e DOCKER_HOST=unix:///docker.sock \
        -v $(pwd):/makisu-context \
        -v /tmp/makisu-storage:/makisu-storage \
        makisu-gcr:latest build --modifyfs=true --load -registry-config=/makisu-context/registry.yaml -t=gcr.io/my-project/christian-playground/makisu --push=gcr.io -f /makisu-context/test.Dockerfile /makisu-context
{"level":"info","ts":1545170766.913012,"msg":"Starting Makisu build (version=v0.1.6-8-g2312b39)"}
{"level":"info","ts":1545170766.9205284,"msg":"Using build context: /makisu-context"}
{"level":"info","ts":1545170766.926709,"msg":"Using file at /makisu-storage/cache_key_value.json for cacheID storage"}
{"level":"info","ts":1545170766.9346206,"msg":"Found mapping in cacheID KVStore: 80dd2397 => b8967b62afd8594745b0b54f713df7b6cf4db380557dd9244ac2113f1af0adb3,db477421f13ba19cb95e4001331af4cb01cb7e85737a12355a9c52a768a4e387"}
{"level":"info","ts":1545170766.9366121,"msg":"* Skipped pulling existing layer my-project/christian-playground/makisu:sha256:db477421f13ba19cb95e4001331af4cb01cb7e85737a12355a9c52a768a4e387"}
{"level":"info","ts":1545170766.9366856,"msg":"* Stage 1/1 : (alias=0,latestfetched=-1)"}
{"level":"info","ts":1545170766.936716,"msg":"* Step 1/2 (commit,modifyfs) : FROM alpine  (76c4d71f)"}
{"level":"info","ts":1545170766.9368548,"msg":"* Started pulling image index.docker.io/library/alpine:latest"}
{"level":"info","ts":1545170768.4722087,"msg":"* Skipped pulling existing layer library/alpine:sha256:4fe2ade4980c2dda4fc95858ebb981489baec8c1e4bd282ab1c3560be8ff9bde"}
{"level":"info","ts":1545170768.4731565,"msg":"* Skipped pulling existing layer library/alpine:sha256:196d12cf6ab19273823e700516e98eb1910b03b17840f9d5509f03858484d321"}
{"level":"info","ts":1545170768.4753537,"msg":"* Finished pulling image index.docker.io/library/alpine:latest in 1.538440486s"}
{"level":"info","ts":1545170768.487996,"msg":"* Processing FROM layer 4fe2ade4980c2dda4fc95858ebb981489baec8c1e4bd282ab1c3560be8ff9bde"}
{"level":"info","ts":1545170768.5707114,"msg":"* Untarred 469 files to / in 82ms"}
{"level":"info","ts":1545170768.571047,"msg":"* Merged 469 headers from tar to memfs"}
{"level":"info","ts":1545170768.57108,"msg":"* Execute FROM alpine  (76c4d71f) took 1.634346367s"}
{"level":"info","ts":1545170768.573798,"msg":"* Committed gzipped layer sha256:4fe2ade4980c2dda4fc95858ebb981489baec8c1e4bd282ab1c3560be8ff9bde (2206931 bytes)"}
{"level":"info","ts":1545170768.5738807,"msg":"* Pushing with cache ID 76c4d71f"}
{"level":"info","ts":1545170768.5755074,"msg":"* Step 2/2 (commit,modifyfs) : RUN echo foo >> bar.txt  (80dd2397)"}
{"level":"info","ts":1545170768.584127,"msg":"* Applying cache layer db477421f13ba19cb95e4001331af4cb01cb7e85737a12355a9c52a768a4e387 (unpack=true)"}
{"level":"info","ts":1545170768.5845675,"msg":"* Untarred 3 files to / in 0s"}
{"level":"info","ts":1545170768.5846548,"msg":"* Merged 3 headers from tar to memfs"}
{"level":"info","ts":1545170768.5847301,"msg":"* Skipping execution; cache was applied *"}
{"level":"info","ts":1545170768.5851738,"msg":"* Moving directories [] to /makisu-storage/sandbox/sandbox154813431/stages/MA=="}
{"level":"error","ts":1545170768.599844,"msg":"Failed to push cache: push layer sha256:4fe2ade4980c2dda4fc95858ebb981489baec8c1e4bd282ab1c3560be8ff9bde: check layer exists: gcr.io/my-project/christian-playground/makisu (sha256:4fe2ade4980c2dda4fc95858ebb981489baec8c1e4bd282ab1c3560be8ff9bde): get security opt: get credentials: get credentials from helper gcr: error getting credentials - err: fork/exec /makisu-internal/docker-credential-gcr: no such file or directory, out: ``"}
{"level":"info","ts":1545170768.6192143,"msg":"Computed total image size 2207126","total_image_size":2207126}
{"level":"info","ts":1545170768.6200254,"msg":"Successfully built image my-project/christian-playground/makisu:latest"}
{"level":"info","ts":1545170768.6203363,"msg":"* Started pushing image gcr.io/my-project/christian-playground/makisu:latest"}
2018/12/18 22:06:08 Command failure: failed to push image: failed to push image: check manifest exists for image gcr.io/my-project/christian-playground/makisu:latest: get security opt: get credentials: get credentials from helper gcr: error getting credentials - err: fork/exec /makisu-internal/docker-credential-gcr: no such file or directory, out: ``
```

So basically it reports that `/makisu-internal/docker-credential-gcr` does not exist which is not true. It's definately in the image at this path. I tried debugging this is a bit and it seems makisu is doing some funky operations to the /makisu-internal directory and somehow deletes / moves around `docker-credential-gcr` during execution. Could you please look into this? Thanks

**Comment by apourchet** [comment]
@geekflyer I've run into this weird error message a bunch of times in the past, and figured it was because the `docker-credential-gcr` binary needs to be statically compiled. So for the makisu-gcr image try using this dockerfile:
```
FROM golang:1.11 AS builder

RUN mkdir -p /go/src/github.com/uber/makisu
WORKDIR /go/src/github.com/uber/makisu

ADD Makefile .
RUN make ext-tools/Linux/dep

ADD Gopkg.toml Gopkg.lock ./
ADD .git ./.git
ADD cli ./cli
ADD bin ./bin
ADD lib ./lib
RUN make bins

# Build GCR credential helper
RUN go get -u github.com/GoogleCloudPlatform/docker-credential-gcr
RUN cd /go/src/github.com/GoogleCloudPlatform/docker-credential-gcr && \
    CGO_ENABLED=0 GOOS=linux go build -tags bins -gcflags '-N -l' -o /docker-credential-gcr main.go

FROM scratch
COPY --from=builder /go/src/github.com/uber/makisu/bin/makisu/makisu /makisu-internal/makisu
COPY --from=builder /docker-credential-gcr /makisu-internal/docker-credential-gcr
ADD ./assets/cacerts.pem /makisu-internal/certs/cacerts.pem
ENTRYPOINT ["/makisu-internal/makisu"]
```
Now we get this new error message:
```
{"level":"info","ts":1545173640.4949303,"msg":"* Started pushing image gcr.io/my-project/christian-playground/makisu:latest"}
2018/12/18 22:54:00 Command failure: failed to push image: failed to push image: check manifest exists for image gcr.io/my-project/christian-playground/makisu:latest: get security opt: get credentials: get credentials from helper gcr: error getting credentials - err: exit status 1, out: `docker-credential-gcr/helper: could not retrieve GCR's access token: google: could not find default credentials. See https://developers.google.com/accounts/docs/application-default-credentials for more information.`
```

**Comment by geekflyer** [comment]
ok thanks @apourchet. I'm getting the same "new" error message as you now. One other thing I noticed that makisu seems to somehow mess with some environment variables / working directory or so.

I can actually confirm that the credential helper works inside the container via CLI invocations by running:
```bash
echo "gcr.io" | docker run -i --rm --network host \
        -v /var/run/docker.sock:/docker.sock \
        -v $HOME/.config/gcloud:/root/.config/gcloud -e DOCKER_HOST=unix:///docker.sock \
        -v $(pwd):/makisu-context \
        -v /tmp/makisu-storage:/makisu-storage \
        --entrypoint "/makisu-internal/docker-credential-gcr" makisu-gcr:latest get
```

which returns valid credentials / an access token for GCR.
However since makisu seems to override some context / env vars the credential helper doesn't work properly when invoked via makisu. I'm guessing it might have to do how (if at all) makisu passes the `HOME` environment variable to the subshell. The credential helper needs the HOME environment variable since it's attempting to load the application_default_credentials.json from `$HOME/.config/gcloud/application_default_credentials.json` by default.

**Comment by evelynl94** [comment]
https://github.com/docker/docker-credential-helpers/blob/5241b46610f2491efdf9d1c85f1ddf5b02f6d962/client/command.go#L36 shows it appends to os.Environ()

According to https://github.com/GoogleCloudPlatform/docker-credential-gcr#gcr-credentials, maybe setting GOOGLE_APPLICATION_CREDENTIALS might also work?

**Comment by evelynl94** [comment]
okay another environment variable we need is `SSL_CERT_DIR`. I got this working with one of our private images. My command:

`$ docker run -i --rm --network host -e SSL_CERT_DIR=/makisu-internal/certs/ -e GOOGLE_APPLICATION_CREDENTIALS=/root/.config/gcloud/application_default_credentials.json -v $HOME/.config/gcloud:/root/.config/gcloud -v $(pwd):/makisu-context -v /tmp/makisu-storage:/makisu-storage makisu:v0.1.6-8-g2312b39 build -t makisu-test:920215  -build-args {} -modifyfs=true -commit=explicit --registry-config /makisu-context/registry.yaml -cache-ttl 0 /makisu-context`

**Comment by geekflyer** [comment]
ok awesome. I got it working now using this dockerfile:

```
FROM golang:1.11 AS builder

RUN mkdir -p /go/src/github.com/uber/makisu
WORKDIR /go/src/github.com/uber/makisu

ADD Makefile .
RUN make ext-tools/Linux/dep

ADD Gopkg.toml Gopkg.lock ./
ADD .git ./.git
ADD cli ./cli
ADD bin ./bin
ADD lib ./lib
RUN make bins

# Build GCR credential helper
RUN go get -u github.com/GoogleCloudPlatform/docker-credential-gcr
RUN cd /go/src/github.com/GoogleCloudPlatform/docker-credential-gcr && \
    CGO_ENABLED=0 GOOS=linux go build -tags bins -gcflags '-N -l' -o /docker-credential-gcr main.go

FROM scratch
COPY --from=builder /go/src/github.com/uber/makisu/bin/makisu/makisu /makisu-internal/makisu
COPY --from=builder /docker-credential-gcr /makisu-internal/docker-credential-gcr
ADD ./assets/cacerts.pem /makisu-internal/certs/cacerts.pem
ENV SSL_CERT_DIR=/makisu-internal/certs
ENV HOME=/root
ENTRYPOINT ["/makisu-internal/makisu"]
```

and this build command:
```
docker run -i --rm --network host \
        -v /var/run/docker.sock:/docker.sock \
        -v $HOME/.config/gcloud:/root/.config/gcloud -e DOCKER_HOST=unix:///docker.sock \
        -v $(pwd):/makisu-context \
        -v /tmp/makisu-storage:/makisu-storage \
        makisu-gcr:latest build --modifyfs=true --load -registry-config=/makisu-context/registry.yaml -t=gcr.io/my-project/christian-playground/makisu --push=gcr.io -f /makisu-context/test.Dockerfile ./makisu-context
```

I actually found that the `scratch` image doesn't come set with a `HOME` environment variable by default which caused the credential helper not to find the credentials, so I'm setting the HOME environment variable to `/root` now in the image build.

**Comment by evelynl94** [comment]
I think we are going to add `docker-credential-gcr` and `amazon-ecr-credential-helper` in our next release image, but Im hesitant to put these environment variables in our Dockerfile...they are pretty random ipo. However, the build command would look messier...

Thoughts? @yiranwang52 @apourchet

**Comment by yiranwang52** [comment]
Added some documentation for this in #153
